### PR TITLE
3D-3D Registration based on PCL

### PR DESCRIPTION
**Pull request by fabiencastan**
WARNING: DO NOT MERGE due to Eigen alignment flags or C++17 switch.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -340,12 +341,10 @@ endif()
 # - external if FLANN_INCLUDE_DIR_HINTS and a valid Flann setup is found
 # ==============================================================================
 if(ALICEVISION_BUILD_SFM)
-  if(NOT DEFINED FLANN_INCLUDE_DIR_HINTS)
-    set(FLANN_INCLUDE_DIR_HINTS ${CMAKE_CURRENT_SOURCE_DIR}/dependencies/flann/src/cpp)
-    set(ALICEVISION_USE_INTERNAL_FLANN ON)
-  endif()
   find_package(Flann QUIET)
   if(NOT FLANN_FOUND OR ALICEVISION_USE_INTERNAL_FLANN)
+    set(FLANN_INCLUDE_DIR_HINTS ${CMAKE_CURRENT_SOURCE_DIR}/dependencies/flann/src/cpp)
```
Comment (gregoire-dl) [inline src/CMakeLists.txt]: `FLANN_INCLUDE_DIR_HINTS` is now set after `find_package`.
In order to find the internal one we should add another `find_package` after this line.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -320,8 +321,8 @@ if(Eigen3_FOUND OR EIGEN3_FOUND)
   # message(FATAL_ERROR "EIGEN_INCLUDE_DIR: ${EIGEN_INCLUDE_DIR}")
   include_directories(${EIGEN_INCLUDE_DIR})
   # See https://eigen.tuxfamily.org/dox/group__TopicUnalignedArrayAssert.html
-  add_definitions("-DEIGEN_DONT_ALIGN_STATICALLY=1")
-  add_definitions("-DEIGEN_DONT_VECTORIZE=1")
+#  add_definitions("-DEIGEN_DONT_ALIGN_STATICALLY=1")
```
Comment (fabiencastan) [inline src/CMakeLists.txt]: WARNING: do not merge before we have done all the changes needed for that.

File: src/aliceVision/registration/PointcloudRegistration.cpp [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+  duration.loadSourceCloud = std::chrono::duration <double, std::milli> (std::chrono::steady_clock::now()-tic).count();
+
+  return res;
+}
+
+int PointcloudRegistration::loadTargetCloud(const std::string & file)
+{
+  auto tic = std::chrono::steady_clock::now();
+
+  int res = loadCloud(file, targetCloud);
```
Comment (simogasp) [inline src/aliceVision/registration/PointcloudRegistration.cpp:55]: const

File: src/aliceVision/registration/PointcloudRegistration.cpp [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+  int res = loadCloud(file, sourceCloud);
+
+  duration.loadSourceCloud = std::chrono::duration <double, std::milli> (std::chrono::steady_clock::now()-tic).count();
+
+  return res;
+}
+
+int PointcloudRegistration::loadTargetCloud(const std::string & file)
+{
+  auto tic = std::chrono::steady_clock::now();
```
Comment (simogasp) [inline src/aliceVision/registration/PointcloudRegistration.cpp:53]: use Timer

File: src/aliceVision/registration/PointcloudRegistration.cpp [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+{
+  if (!boost::filesystem::exists(inputFile))
+  {
+    ALICEVISION_LOG_ERROR("PointcloudRegistration::saveCloud: The file does not exist '" << inputFile << "'");
+    return EXIT_FAILURE;
+  }
+
+  pcl::PointCloud<pcl::PointXYZRGB> inputCloud, outputCloud;
+
+  // load input cloud:
```
Comment (simogasp) [inline src/aliceVision/registration/PointcloudRegistration.cpp:75]:
```c++
const auto ext =  path(inputFile).extension().toString();
if(inputFile == "ply")
...
```

File: src/aliceVision/registration/PointcloudRegistration.cpp [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+  {
+    ALICEVISION_LOG_ERROR("PointcloudRegistration::saveCloud: Unknown extension: " << inputFile);
+    return EXIT_FAILURE;
+  }
+
+  // transform input pointcloud according to T:
+  pcl::transformPointCloud(inputCloud, outputCloud, T);
+
+  // save the transformed pointcloud:
+  if (outputFile.substr(outputFile.find_last_of(".") + 1) == "ply")
```
Comment (simogasp) [inline src/aliceVision/registration/PointcloudRegistration.cpp:98]:
```c++
if(ext != "ply")
{
    ALICEVISION_LOG_ERROR("PointcloudRegistration::saveCloud: Unknown extension: " << outputFile);
    return EXIT_FAILURE;
}
pcl::io::savePLYFile(outputFile, outputCloud);
```

**Review by simogasp — CHANGES_REQUESTED** [review]

**Comment by fabiencastan** [comment]
Check setUseSymmetricObjective:  https://github.com/PointCloudLibrary/pcl/blob/49385a6f9f37c9a1b4e4fa5f6e5b2019b90ac4b8/registration/include/pcl/registration/icp.h#L306

**Comment by griwodz** [comment]
FYI, we have a quick example for using symmetric with PCL > 1.10.0 here:
https://github.com/henryhansen23/Indoor-reconstruction-plus-plus/pull/3/files

File: src/aliceVision/registration/PointcloudRegistration.cpp [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+    ALICEVISION_LOG_INFO("|- T_origin_target = \n" << targetInitialTransform);
+
+    Eigen::Matrix4d Ts = preparePointClouds(mutableSourceCloud, mutableTargetCloud);
+
+    Eigen::Matrix4d transform;
+    switch (mode)
+    {
+        case EAlignmentMethod::GICP:
+        {
+            transform = alignGICP(mutableSourceCloud, mutableTargetCloud);
```
Comment (simogasp) [inline src/aliceVision/registration/PointcloudRegistration.cpp:197]: `break`? 
or is it done on purpose?

Comment (fabiencastan) [reply src/aliceVision/registration/PointcloudRegistration.cpp:197]: it seems to be a bug

**Pull request closed without merging**